### ISSUE TITLE
Deleting Annotations

### ISSUE THREAD

**Issue by cmbirk** [opengovfoundation/madison#122]
@krues8dr and @seamuskraft ,
I'm torn with what to do when a user deletes an annotation ( they can only delete their own ).  I think users should be able to delete their annotations, but if their annotation has say - 100 comments on it - those comments disappear as well.  Are we ok with this?  Should the annotation get archived instead ( we would have to update our terms & conditions for that )?  If we're going with archived, only save them for a certain amount of time?

**Comment by krues8dr** [comment]
Bumping this.  Users are asking for this feature now.

**Comment by cmbirk** [comment]
@stvnrlly would love your feedback on the thread above.

@krues8dr brought up another good point this morning, which is that an issue that reddit has seen is someone getting the top post on an article and then changing the content to an advertisement.

I think the 15 minute window seems like a good compromise here.  Anything after that would require another comment clarifying the original post.

**Comment by krues8dr** [comment]
Pinging @mattbailey0 as well, since this will impact them directly.

**Comment by seamuskraft** [comment]
Deleting annotations or proposed edits is a big policy decision, for core and for DC's instance. What DC wants to do on their instance should be considered separately from what goes into core Madison.  In addition, the DC instance has to take into account the DC government's public records retention laws/policies.  Might be good to loop in Dave Zvenyach, too.

**Comment by krues8dr** [comment]
Paging @vzvenyach :bell:

**Comment by vzvenyach** [comment]
So, there may be a middle ground. Perhaps the right strategy is not to *delete* the annotation, but instead to create an archive flag where the the annotation is not discoverable through search or shows up on the page, but the permanent url persists.

**Comment by seamuskraft** [comment]
@vzvenyach Thank you sir. that could work. Users could also have the ability to edit their responses (for a given time period...an hour? 15 minutes?) but it a) shows on the annotation or comment that it has been edited and b) if you click on the "edited" or something to that effect, you can see the original comment/suggestion/annotation.

**Comment by cmbirk** [comment]
Well right now the link to an annotation is to the anchor on the page.  We don't have separate pages for annotations.

**Comment by krues8dr** [comment]
@cmbirk except for the RSS feed, of course.

@seamuskraft @vzvenyach I like this suggestion.  On the dev side, it's an order of magnitude more complicated to keep track of revisions, but that shouldn't keep us from doing it. (It just might keep us from doing it _today_ ;) )

**Comment by cmbirk** [comment]
@krues8dr so the annotation would still exist in the RSS feed if deleted?

How would a user that made a comment on the archived annotation know where to find the original URL?

**Comment by krues8dr** [comment]
Dunno.  Just throwing it out there as a complication. :)

**Comment by stvnrlly** [comment]
Archiving sounds good to me, but I also think there should be an `undo` option for whatever the user _just_ posted. That makes contributing feel less like a high-wire act.

**Comment by seamuskraft** [comment]
@krues8dr yeah, I much much prefer the "keep us from doing it today" approach, not just because that suggestion would be complicated to execute.  Since there are a number of unknowns here, and policy/legal/user control/user comfort implications here to think through, that might be the best way to go anyway.

**Comment by cmbirk** [comment]
@seamuskraft yeah, we aren't going to solve this issue today, but did need to bump the conversation.

If we're going the archiving route, we're going to need to create annotation permalinks (that don't lose context with the document) and a way to allow users to find their sub-content that has been archived.

**Comment by seamuskraft** [comment]
@cmbirk we absolutely should have annotation/comment permalinks.  Doesn't Hypothes.is already do this/have this feature?

**Comment by seamuskraft** [comment]
maybe that's the feature to tackle while we game out the larger questions at play here.

**Comment by cmbirk** [comment]
We have permalinks to annotations in the page.  This keeps the context of the annotation in the document.  Just having a page with the annotation text and selection isn't particularly useful, but is good for data persistence.  

That will be likely what we'll need to create, but I would still have the user copy the permalink to the annotation in the document context.  That means we still need a way to allow the user to discover / find archived sub-content.

**Comment by rebeccawilliams** [comment]
FWIW, you can edit your comment on other annotation type platforms:
* Google Docs
* GitHub 

Also, FWIW, I had to write this comment twice and now it's more confusing for everyone:
* https://drafts.dc.gov/docs/draft-open-data-policy#annotation_185 whoops I hit post
* https://drafts.dc.gov/docs/draft-open-data-policy#annsubcomment_95 whoops I added this as a comment to another person's annotation trying to fix the first post and now I can't move it 

I am pro-Let Users Edit Their Comments :+1:

**Comment by emanuelfeld** [comment]
I have flagged a separate, but I think related, concern here https://github.com/DCgov/dc-madison/issues/62 regarding sponsor/submitters updating their document. How should (could) annotations from previous versions be reflected in updates?

**Comment by emanuelfeld** [comment]
To stem the accidental submissions, might also just remove the 'enter to submit' functionality.

**Comment by krues8dr** [comment]
It sounds like we have a few things on the table, I'll try to summarize so we can get actionable items - I am using comments and annotations interchangeably in the following :

In short the problem is:

* Users want to edit, and delete, comments and annotations.
* Policies or laws may require that all submitted comments stay part of the public record, and cannot be deleted.  
* There is probably not a one-size-fits-all solution for everyone who's standing up Madison.

The solutions proposed (policy-related):

1) Allow a 15 minute window where things can be edit & deleted, afterwards they're permanently in place.  (I think this also preserves @stvnrlly 's `undo` idea? Please tell me if not.)
2) Always allow users to edit & delete, but keep a publicly-viewable record of edits, and deletes are hidden by default but still viewable (clicking a toggle to show or otherwise)
3) Always allow users to edit & delete, but the edits and deletes are not publicly viewable, but remain in the database.  (With some method, I assume, for admins to retrieve those for public records requests.)

And also, technically:

4) Remove the submit-on-enter-keypress to reduce accidental submissions.

And I would add:

5) Do not allow users to submit comments that contain only punctuation and/or spaces - I saw a few of these in the database while exporting.

(I assume it goes without saying that 4) is a great idea and we should do it regardless.  Thanks for that @emanuelfeld ! )

A related problem is that admins need to be able to do _something_ with flagged comments (or otherwise).  Internally we'd discussed not allowing admins to delete comments, but only to hide them temporarily, similar to 2) above.


Since we need a solution for DC, we need to know authoritatively ( @mattbailey0 @vzvenyach ) :

a) From a legal standpoint, are these comments considered part of the public record and must be preserved?
b) Would any of the options above go against public record laws as they are currently written, or could they be interpreted that way?

Absent that knowledge, my gut says that 2) is the safest and the most applicable to the most use cases, not just for DC.  It's non-destructive, it preserves information publicly and transparently, and gives users an easy option.  It also behaves _similarly_ to other platforms such as reddit and facebook.

To @seamuskraft 's point - this isn't something to be taken lightly, and we might not be able to come up with a single solution for all cases.  However, I feel that the `perfect is the enemy of the good (enough)`. Implementing a solution that works for DC now, and adding other options down the road as we encounter new use cases, is a very reasonable way to address this.  

If we can sort out the legal question, I think we are in a good enough place that we can move ahead with one of the options listed above.  Unless there are other options I haven't listed or we haven't considered, let's try and move this towards a single resolution for the immediate need.

**Comment by stvnrlly** [comment]
Re: `undo`: I was imagining this in addition to a tracked edit/delete option. Something like a 30-second grace period in which hitting the `undo` button puts you back into the "add annotation" bubble with the text that you had entered. 

(This would be less necessary, but still useful, after https://github.com/opengovfoundation/madison/issues/805)

**Comment by krues8dr** [comment]
Thanks @stvnrlly I've updated the options!
